csharp/src/Microsoft.ML.OnnxRuntime/NativeMethods.cs

```diff
@@ -260,7 +260,7 @@ static NativeMethods()
                                                 string[] outputNames,
                                                 UIntPtr outputCount,
 
-                                                [MarshalAs(UnmanagedType.LPArray, SizeParamIndex = 5 /*index of outputCount*/)][In, Out]
+                                                [MarshalAs(UnmanagedType.LPArray, SizeParamIndex = 6 /*index of outputCount*/)][In, Out]
```

Comment:
IMO we should just remove the `MarshalAs` attribute all together. It is not helping anything.

This array is always created in managed code:

https://github.com/microsoft/onnxruntime/blob/034aa8016767db946887d2ae53b05ddee76592d3/csharp/src/Microsoft.ML.OnnxRuntime/InferenceSession.cs#L142-L152

And it is never changed in native code. It is only read, and filled with pointers.

https://github.com/microsoft/onnxruntime/blob/686bd36210affca89d3791e696e84943a57d2f7b/onnxruntime/core/session/onnxruntime_c_api.cc#L417-L421

https://github.com/microsoft/onnxruntime/blob/686bd36210affca89d3791e696e84943a57d2f7b/onnxruntime/core/session/onnxruntime_c_api.cc#L438-L440

So if we just remove `MarshalAs` altogether, this will just work.

/cc @AaronRobinsonMSFT

Reply:
We have this same pattern in corefx:

https://github.com/dotnet/corefx/blob/b49a8a9be1d53cd9e50cb68fd8540be25c65d433/src/System.Drawing.Common/src/System/Drawing/GdiplusNative.cs#L445-L446

https://github.com/dotnet/corefx/blob/dc7005cd64787880dfc27f19120944d46fb8f200/src/System.Drawing.Common/src/System/Drawing/Text/FontCollection.cs#L43-L46

And we don't need `MarhsalAs` at all, and it works on all supported platforms.

Note that if this was truly an `Out` value where the C code was returning an array created in the native code, and we wanted to marshal it to a .NET array, then you would use the `MarshalAs` attribute. But that isn't the case here.

Reply:
@eerhardt Is giving some very solid advice. Unfortunately this kind of marshalling is error prone and can exist for a long time prior to causing an error.

Reply:
The outputValueArray  is InOut.

Reply:
Ort Runtime may read the array's contents. Each element of the array should be either null, or a **valid** OrtValue pointer.

Reply:
@snnn - With the current code, the reference to the array is not an `InOut`, it is only ever an `In`. The C code might be changing values inside the array, but it isn't changing which array the reference is pointing to.

Reply:
The Array itself is not going to be reallocated, and the outputSize is not an InOut ptr, so I think it is safe to remove the MarshalAs. @eerhardt, I assume the default marshalling of IntPtr array will be a pinned reference.

Reply:
Yes, the default marshaling will pin the array while control is inside the native method.